src/compiler/checker.ts

```diff
@@ -25338,10 +25338,25 @@ namespace ts {
             // control flow analysis an opportunity to narrow it further. For example, for a reference of a type
             // parameter type 'T extends string | undefined' with a contextual type 'string', we substitute
             // 'string | undefined' to give control flow analysis the opportunity to narrow to type 'string'.
-            const substituteConstraints = !(checkMode && checkMode & CheckMode.Inferential) &&
-                someType(type, isGenericTypeWithUnionConstraint) &&
-                (isConstraintPosition(type, reference) || hasContextualTypeWithNoGenericTypes(reference, checkMode));
-            return substituteConstraints ? mapType(type, t => t.flags & TypeFlags.Instantiable ? getBaseConstraintOrType(t) : t) : type;
+            if (checkMode && checkMode & CheckMode.Inferential) {
+                return type;
+            }
+
+            let contextualType: Type | undefined;
+            // If we aren't in a constraint position, or we can't find a contextual type, or the contextual type indicates
+            // that the type in question may be a direct inference source, then don't do anything special.
+            if (!isConstraintPosition(type, reference) && !(contextualType = tryGetContextualTypeWithNoGenericTypes(reference, checkMode))) {

```

Comment:
I'm pretty sure the slowdown comes from fetching the contextual type way more often. Previously we'd only do it when `someType(type, isGenericTypeWithUnionConstraint)` is true, which is fairly rare.

Reply:
I'd suggest adding a check that `someType(type, t => !!(t.flags & TypeFlags.Instantiable))`, and bail out if not _before_ fetching the contextual type. That would reduce the overhead to only cases where something is generic.